test/libsolidity/syntaxTests/parsing/single_return_param_trailing_comma.sol

```diff
@@ -3,3 +3,4 @@ contract test {
 }
 // ----
 // ParserError: (43-44): Unexpected trailing comma in parameter list.
+// ParserError: (49-50): Expected pragma, import directive or contract/interface/library definition.
```

Comment:
Is there a simple way we can make this result in exactly the same AST as
```
contract test {
	function() returns (uint a) {}
}
```
without the additional error?

Reply:
Copying from what I wrote in the gitter channel:

The problem is that you don't know the intention of the programmer.  Of course we can drive this down and some mechanisim has been put in place for that. Such as noting that we are inside parser error recovery.

If  _ExpectToken()_ is changed so that it returns whether it succeeded or not then calling routines could then not gobble up important symbols which help in error recovery.

My take is to start out with the additional error messages. I can create a todo list of how to reduce errors caused by (less optimial) error recovery and then come back to that.

We can go down a big rabbit hole with this. If you want something simple, the Steve C Johnson approach is the simplest and most general there is. You'll have to live with a little bit of slop though.

So changing the check in error reporting may be the most expedient thing for now. 

But please let's wait until we get the fuller bigger picture done first. For IDE's it is helpful to have more information not less. That doesn't mean the end IDE user needs to see this, just that it is available for an IDE to make use of.

Reply:
Maybe the contract level is not the best way to recover as a first step. It searches for a `}`, but that token is used all over the place in the grammar. Wouldn't it be better to first recover at statement level because whenever we see a `;`, we can be sure that we are at statement level.

Reply:
I don't understand what you would be suggesting.

We _do_ try to recover at  a statement level. However in 

```
contract test {
	function() returns (uint a,) {}
}
```
there _is_ no semicolon. In fact we are not looking for a statement here. The rule nesting looks like this:
```
   ContractDefinition ->
   FunctionDefinitionOrFunctionTypeStateVariable ->
   FunctionHeader ->
   ParameterList
```

Of the list of `Block`, `ContractDefinition`, `Statement` the only thing available is `ContractDefinition`.

Reply:
What I'm suggesting is to not recover errors outside the statement level at all because it will worsen the error reports.

Reply:
If there isn't recovery at ContractDefinition then an IDE has nothing to go on here in this particular example.

Not even that there was a contract named `test`.  Furthermore, if there happen to be multiple contracts and libraries, once one fails, they all fail. 

is that what is desired?

Reply:
If there is a fatal error outside of the statement scope, then the whole file fails, yes. If there is a fatal error inside statement scope, everything is fine. I would say this is a good compromise given we have a production-ready software here that we want to extend with an experimental feature.

Reply:
A similar idea -  how about if we add an option for extended error recovery which is set off by default.

With this, none (or almost none) of the existing 600 or so tests need to be modified. However those front-end tools that would like error recovery would still be able to get more information.

Reply:
That sounds like a very good idea!

Reply:
Excellent! I'll do that now.

Reply:
Done now in 0e6568cc5